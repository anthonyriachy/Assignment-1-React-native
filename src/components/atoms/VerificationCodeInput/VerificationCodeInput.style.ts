import { StyleSheet } from "react-native";
export const createStyles = (colors: any) => StyleSheet.create({
    container: {
        flexDirection: 'row',
        justifyContent: 'space-between',
        width: '100%',
        gap:10,
    },
    input: {
        width: 50,
        height: 50,
        borderRadius: 8,
        textAlign: 'center',
        fontSize: 24,
        backgroundColor: colors.inputBackground,
        color: colors.text,
    },
}); 