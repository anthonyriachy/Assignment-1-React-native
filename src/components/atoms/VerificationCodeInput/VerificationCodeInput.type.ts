export type VerificationCodeInputProps = {
    code: string[];
    setCode: (code: string[]) => void;
}