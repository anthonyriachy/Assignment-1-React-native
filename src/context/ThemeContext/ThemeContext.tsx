import React, { createContext, useEffect, useState } from 'react';
import { useColorScheme } from 'react-native';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { ThemeType, getThemeColors } from '../../constants/theme';
import { ThemeContextType } from './ThemeContext.type';

export const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

export const ThemeProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  const systemColorScheme = useColorScheme();
  const [theme, setTheme] = useState<ThemeType>(systemColorScheme || 'light');

  useEffect(() => {
    loadTheme();
  }, []);

  const loadTheme = async () => {
    try {
      const savedTheme = await AsyncStorage.getItem('theme');
      if (savedTheme) {
        setTheme(savedTheme as ThemeType);
      }
    } catch (error) {
      console.error('Error loading theme:', error);
    }
  };

  const toggleTheme = () => {
    const newTheme = theme === 'light' ? 'dark' : 'light';
    setTheme(newTheme);
    AsyncStorage.setItem('theme', newTheme);
  };

  const value = {
    theme,
    colors: getThemeColors(theme),
    toggleTheme,
    setTheme: (newTheme: ThemeType) => {
      setTheme(newTheme);
      AsyncStorage.setItem('theme', newTheme);
    },
  };

  return <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>;
};

