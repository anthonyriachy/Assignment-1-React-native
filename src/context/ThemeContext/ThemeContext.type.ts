import { getThemeColors, ThemeType } from '../../constants/theme';

export type ThemeContextType = {
    theme: ThemeType;
    colors: ReturnType<typeof getThemeColors>;
    toggleTheme: () => void;
    setTheme: (theme: ThemeType) => void;
  }