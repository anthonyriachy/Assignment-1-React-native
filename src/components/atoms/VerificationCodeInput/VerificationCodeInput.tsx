import React, { useRef } from 'react';
import { View, TextInput } from 'react-native';
import { createStyles } from './VerificationCodeInput.style';
import { useTheme } from '../../../hooks/UseTheme';

interface VerificationCodeInputProps {
    onCodeComplete: (code: string) => void; 
    code: string[];
    setCode: (code: string[]) => void;
}

export const VerificationCodeInput = ({ code, setCode }: VerificationCodeInputProps) => {
    const { colors } = useTheme();
    const styles = createStyles(colors);
    const inputRefs = useRef<Array<TextInput | null>>([]);

    const handleCodeChange = (text: string, index: number) => {
        if(isNaN(Number(text))) return;
        const newCode = [...code];
        newCode[index] = text;
        setCode(newCode);

        
        if (text && index < 3) {
            inputRefs.current[index + 1]?.focus();
        }
    };

    return (
        <View style={styles.container}>
            {[0, 1, 2, 3].map((index) => (
                <TextInput
                    key={index}
                    ref={(ref) => {
                        inputRefs.current[index] = ref;
                    }}
                    style={styles.input}
                    maxLength={1}
                    keyboardType="number-pad"
                    value={code[index]}
                    onChangeText={(text) => handleCodeChange(text, index)}
                    // onKeyPress={(e) => handleKeyPress(e, index)}
                />
            ))}
        </View>
    );
};
 