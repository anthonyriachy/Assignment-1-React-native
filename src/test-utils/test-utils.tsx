import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider } from '../context/ThemeContext/ThemeContext';

// Mock theme hook
jest.mock('../hooks/UseTheme', () => ({
  useTheme: () => ({
    colors: {
      background: '#fff',
      text: '#000',
      primary: '#007AFF',
      border: '#ccc',
    },
    theme: 'light',
  }),
}));

// Custom render function that includes providers
const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: ThemeProvider, ...options });

// Re-export everything
export * from '@testing-library/react-native';

// Override render method
export { customRender as render }; 