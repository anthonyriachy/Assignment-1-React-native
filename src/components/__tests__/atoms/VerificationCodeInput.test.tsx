import React from 'react';
import { fireEvent } from '@testing-library/react-native';
import { render } from '../../../../src/test-utils/test-utils';
import { VerificationCodeInput } from '../../atoms/VerificationCodeInput/VerificationCodeInput';
import { ThemeProvider } from '../../../context/ThemeContext/ThemeContext';

// Mock useTheme hook
jest.mock('../../../hooks/UseTheme', () => ({
  useTheme: () => ({
    colors: {
      background: '#fff',
      text: '#000',
      primary: '#007AFF',
      border: '#ccc',
    },
  }),
}));

describe('VerificationCodeInput', () => {
  const initialCode = ['', '', '', '', '', ''];

  it('renders correctly with empty code', () => {
    const { toJSON } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={initialCode}
          setCode={jest.fn()}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with partial code', () => {
    const partialCode = ['1', '2', '3', '', '', ''];
    const { toJSON } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={partialCode}
          setCode={jest.fn()}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly with complete code', () => {
    const completeCode = ['1', '2', '3', '4', '5', '6'];
    const { toJSON } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={completeCode}
          setCode={jest.fn()}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders 6 input fields', () => {
    const { getAllByTestId } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={initialCode}
          setCode={jest.fn()}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );

    const inputs = getAllByTestId(/verification-input-\d/);
    expect(inputs).toHaveLength(6);
  });

  it('handles numeric input correctly', () => {
    const mockSetCode = jest.fn();
    const { getAllByTestId } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={initialCode}
          setCode={mockSetCode}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );

    const inputs = getAllByTestId(/verification-input-\d/);
    fireEvent.changeText(inputs[0], '1');
    
    expect(mockSetCode).toHaveBeenCalledWith(['1', '', '', '', '', '']);
  });

  it('rejects non-numeric input', () => {
    const mockSetCode = jest.fn();
    const { getAllByTestId } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={initialCode}
          setCode={mockSetCode}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );

    const inputs = getAllByTestId(/verification-input-\d/);
    fireEvent.changeText(inputs[0], 'a');
    
    expect(mockSetCode).not.toHaveBeenCalled();
  });

  it('handles input sequence correctly', () => {
    let currentCode = [...initialCode];
    const mockSetCode = jest.fn((newCode) => {
      currentCode = newCode;
    });

    const { getAllByTestId, rerender } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={currentCode}
          setCode={mockSetCode}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );

    const inputs = getAllByTestId(/verification-input-\d/);
    
    // Test each input individually
    fireEvent.changeText(inputs[0], '1');
    expect(mockSetCode).toHaveBeenCalledWith(['1', '', '', '', '', '']);
    
    // Update the component with new state
    rerender(
      <ThemeProvider>
        <VerificationCodeInput 
          code={currentCode}
          setCode={mockSetCode}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );

    fireEvent.changeText(inputs[1], '2');
    expect(mockSetCode).toHaveBeenCalledWith(['1', '2', '', '', '', '']);
    
    // Update the component with new state
    rerender(
      <ThemeProvider>
        <VerificationCodeInput 
          code={currentCode}
          setCode={mockSetCode}
          onCodeComplete={jest.fn()}
        />
      </ThemeProvider>
    );

    fireEvent.changeText(inputs[2], '3');
    expect(mockSetCode).toHaveBeenCalledWith(['1', '2', '3', '', '', '']);
  });

  it('handles complete code entry', () => {
    const mockOnCodeComplete = jest.fn();
    const mockSetCode = jest.fn();
    const { getAllByTestId } = render(
      <ThemeProvider>
        <VerificationCodeInput 
          code={['1', '2', '3', '4', '5', '']}
          setCode={mockSetCode}
          onCodeComplete={mockOnCodeComplete}
        />
      </ThemeProvider>
    );

    const inputs = getAllByTestId(/verification-input-\d/);
    fireEvent.changeText(inputs[5], '6');
    
    expect(mockSetCode).toHaveBeenCalledWith(['1', '2', '3', '4', '5', '6']);
  });
}); 