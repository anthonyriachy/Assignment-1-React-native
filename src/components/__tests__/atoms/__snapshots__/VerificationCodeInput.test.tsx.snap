// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerificationCodeInput renders correctly with complete code 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 10,
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-0"
    value="1"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-1"
    value="2"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-2"
    value="3"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-3"
    value="4"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-4"
    value="5"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-5"
    value="6"
  />
</View>
`;

exports[`VerificationCodeInput renders correctly with empty code 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 10,
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-0"
    value=""
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-1"
    value=""
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-2"
    value=""
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-3"
    value=""
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-4"
    value=""
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-5"
    value=""
  />
</View>
`;

exports[`VerificationCodeInput renders correctly with partial code 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "gap": 10,
      "justifyContent": "space-between",
      "width": "100%",
    }
  }
>
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-0"
    value="1"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-1"
    value="2"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-2"
    value="3"
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-3"
    value=""
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-4"
    value=""
  />
  <TextInput
    keyboardType="number-pad"
    maxLength={1}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": undefined,
        "borderRadius": 8,
        "color": "#000",
        "fontSize": 24,
        "height": 50,
        "textAlign": "center",
        "width": 50,
      }
    }
    testID="verification-input-5"
    value=""
  />
</View>
`;
